server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -335,6 +425,193 @@ public WriteField removeValue(int index) {
         return this;
     }
 
+    /**
+     * Removes the {@param o} if this WriteField contains {@param o} using reference equality.
+     */
+    void remove(Object o) {
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return;
+        }
+
+        if (value instanceof List<?> list) {
+            Iterator<?> it = list.iterator();
+            // List.remove(Object) uses Objects.equals which will check content equality for Maps.
+            while (it.hasNext()) {
+                if (it.next() == o) {
+                    it.remove();
+                    return;
+                }
+            }
+        } else if (container.get(leaf) == value) {
+            container.remove(leaf);
+        }
+    }
+
+    /**
+     * Append a {@link NestedDocument} to this field and return it.
+     */
+    @SuppressWarnings("unchecked")
+    public NestedDocument doc() {
+
+        Object value = get(MISSING);
```

Comment:
I wonder if we can refactor to common some of this code that deals with the different value types between doc() and doc(index)?  I'm thinking in terms of having to maybe support another type in the future or having to fix bugs in both places.

Reply:
I've pulled out the common code for handling types, what do you think?